Source/Particles/Gather/GetExternalFields.cpp

```diff
@@ -0,0 +1,46 @@
+#include "WarpX.H"
+#include "Particles/Gather/GetExternalFields.H"
+
+GetExternalEField::GetExternalEField (const WarpXParIter& a_pti, int a_offset) noexcept
+{
+    auto& warpx = WarpX::GetInstance();
+    auto& mypc = warpx.GetPartContainer();
+    if (mypc.m_E_ext_particle_s=="constant" || mypc.m_E_ext_particle_s=="default")
+    {
+        m_type = Constant;
+        m_field_value[0] = mypc.m_E_external_particle[0];
+        m_field_value[1] = mypc.m_E_external_particle[1];
+        m_field_value[2] = mypc.m_E_external_particle[2];
+    }
+    else if (mypc.m_E_ext_particle_s=="parse_e_ext_particle_function")
+    {
+        m_type = Parser;
+        m_time = warpx.gett_new(a_pti.GetLevel());
+        m_get_position = GetParticlePosition(a_pti, a_offset);
+        m_xfield_partparser = mypc.m_Ex_particle_parser.get();
```

Comment:
Just general thoughts on `m_<x/y/z>field_partparser` in the existing `GetExternalField`: Since this is using a function pointer its probably still doing an indirection. Since the type is the same everywhere this is not too critical but could prevent some inlining and vectorizaton across following calls to it. But in the end it potentially calls the parser anyway, that is doing recursive function calls... So all good, I guess.

Back to the class: Should these members pointers be restricted or can they point to identical parsers?

Reply:
There is definitely some runtime switching going on here and in other places throughout this PR. Ultimately we decided that since many things are runtime parameters in WarpX, it's too cumbersome to have to call a different template specialization for every option, the way we do for `nox` and `lower_order_in_v`. 

I think that, given the indirection, it likely will not make a performance difference whether we use restrict here or not.